src/widgets/reflow.rs

```diff
@@ -72,155 +72,142 @@ where
+                    let mut remaining_width =
+                        u16::saturating_sub(self.max_line_width, current_line_width);
+
+                    wrapped_lines.push(std::mem::take(&mut current_line));
+                    current_line_width = 0;
+
+                    // remove whitespace up to the end of line
+                    while let Some(grapheme) = pending_whitespace.front() {
+                        let width = grapheme.symbol.width() as u16;
+
+                        if width > remaining_width {
+                            break;
                         }
-                        current_line.append(&mut unfinished_word);
-                    }
-                    if !current_line.is_empty() {
-                        wrapped_lines.push(current_line);
+
+                        whitespace_width -= width;
+                        remaining_width -= width;
+                        pending_whitespace.pop_front();
                     }
-                    if wrapped_lines.is_empty() {
-                        // Append empty line if there was nothing to wrap in the first place
-                        wrapped_lines.push(vec![]);
+
+                    // don't count first whitespace toward next word
+                    if is_whitespace && pending_whitespace.is_empty() {
+                        continue;
                     }
+                }
 
-                    self.wrapped_lines = Some(wrapped_lines.into_iter());
+                // append symbol to a pending buffer
+                if is_whitespace {
+                    whitespace_width += symbol_width;
+                    pending_whitespace.push_back(StyledGrapheme { symbol, style });
                 } else {
-                    // No more whole lines available -> stop repeatedly retrieving next wrapped line
-                    break;
+                    word_width += symbol_width;
+                    pending_word.push(StyledGrapheme { symbol, style });
                 }
+
+                non_whitespace_previous = !is_whitespace;
             }
-        }
 
-        if let Some(line) = current_line {
-            self.current_line = line;
-            Some(WrappedLine {
-                line: &self.current_line,
-                width: line_width,
-                alignment: self.current_alignment,
-            })
-        } else {
-            None
+            // append remaining text parts
+            if !pending_word.is_empty() || !pending_whitespace.is_empty() {
+                if current_line.is_empty() && pending_word.is_empty() {
+                    wrapped_lines.push(vec![]);
+                } else if !self.trim || !current_line.is_empty() {
+                    current_line.extend(pending_whitespace);
+                } else {
+                    // TODO: explain why this else branch is ok
+                    // See clippy::else_if_without_else
+                }
+
+                current_line.append(&mut pending_word);
+            }
```

Comment:
This is also a fairly dense if statement - is there a good way to simplify this logic?

Reply:
The simplest way I see is just to wrap the inner if statements in a labeled block with a short circuit if there is no pending word or whitespace.

Reply:
I'm not opposed to writing the same body of the if statements twice here if the conditions are clearer - this can possibly be a longer single level if / else-if / ... statement and be more easy to read and maintain. If this was at the top level in a function, I'd be less annoyed by it. But it's nested within a loop and so far down I have to repeatedly scroll up to work out which block it's at the end of.

I'd avoid using labeled blocks generally - they're niche enough that they make this an already complex function more difficult to read rather than easier.

Reply:
explained better elsewhere.